src/sync/mutex.rs

```diff
@@ -310,6 +310,13 @@ impl<MutexType: RawMutex, T> DerefMut for GenericMutexGuard<'_, MutexType, T> {
     }
 }
 
+// Safety: GenericMutexGuard may only be used across threads if the underlying
+// type is Sync.
+unsafe impl<MutexType: RawMutex, T: Sync> Sync
```

Comment:
Does that really remove the `Sync`ness for other types? Or just add additional `Sync`ness? What do the docs say after this - or did you check that your test then fails to compile?

The other approach I thought about was adding a `PhantomType` of something which isn't `Sync` into `GenericMutexGuard`, but then it's never `Sync`.

Reply:
Yup it overrides the auto implemented `Sync` implementation. The docs now say:

```
impl<MutexType: RawMutex, T: Sync, '_> Sync for GenericMutexGuard<'_, MutexType, T>
```
![image](https://user-images.githubusercontent.com/773529/97794168-c954fa80-1bcc-11eb-857c-5a80fc56610b.png)

I also tested that this simpler reproducer:
```rust
    #[test]
    fn test2() {
        use std::cell::Cell;

        let mutex = Mutex::new(Cell::new(2), true);
        let guard = mutex.try_lock().unwrap();
        is_sync(&guard);
    }
```
with the error:
```
error[E0277]: `Cell<{integer}>` cannot be shared between threads safely
   --> tests/mutex.rs:483:17
    |
451 |     fn is_sync<T: Sync>(_: &T) {}
    |                   ---- required by this bound in `is_sync`
...
483 |         is_sync(&guard);
    |                 ^^^^^^ `Cell<{integer}>` cannot be shared between threads safely
    |
    = help: the trait `Sync` is not implemented for `Cell<{integer}>`
    = note: required because of the requirements on the impl of `Sync` for `GenericMutexGuard<'_, parking_lot::raw_mutex::RawMutex, Cell<{integer}>>`
```